### PR TITLE
Fixed #26429 -- Placed a timestamp into generated merge migration names

### PR DESCRIPTION
**Pull request by rtpg**
This sets a timestamp to the generated merge migration names, much in the same fashion as other migrations. This is to help reduce the possibility of a naming conflict (especially after squashing migrations).

  This is to fix the issues mentioned in [#26469](https://code.djangoproject.com/ticket/26429#comment:3), where the combination of merge migration names and squashing can cause a migration to end up never running due to a name conflict.

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by charettes** [comment]
You should be able to fix the failing tests by mocking `datetime.datetime.now()` to return a fixed date and adjust the expected filenames accordingly.

**Comment by MarkusH** [comment]
I'm wondering if it may make sense to move the file name generation onto the `Migration` class (`django.db.migrations.migration.Migration`). The `MigrationAutodetector` has a method `suggest_name()` that already handles some cases. I feel that would be a more appropriate way to solve the issue.

**Comment by rtpg** [comment]
I had tried another way of fixing this branch (by adding support for `--name` in the `makemigrations` command for merge migrations), but looking at `suggest_name` that seems like the "proper" solution. Plus it helps reduce the amount of command-only code. I'll try moving the name generation code there.

 I felt a bit uneasy at patching `datetime.datetime.now` but thinking about it now, that would be a good way of testing that the default naming scheme works as expected, I'll go with that as well.

**Comment by rtpg** [comment]
I looked into integrating the merge migration naming behaviour into the `MigrationAutodetector` class. It turns out that the autodetector is naming the files, but is mainly being used in the standard migration generator (the one that generates changes with "operations"). Merge migrations aren't really an operation, but just a declaration of dependencies, so it's all being done in a separate ad-hoc area.

The merge migration generation code is actually all within the `makemigrations` command under `handle_merge`, and shares very little with the code path that ends up using the `MigrationAutodetector` class. This is probably not too great; you have to run the command to test the merge migration generator logic. It would be worthwhile to pull out the merge migration handling code into its own module, though I don't know what the Django policy is on "refactor code without any behaviour changes associated". 

Anyways, a long way of saying that I kept the initial way of generating the name because it keeps the diff small without being worse than the status quo. I also added a test that checks that the default naming scheme  works properly through a `datetime

**Comment by rtpg** [comment]
Should I rebase this set of commits? Would doing so mess up the pull request (not too sure  about how well Github's PR system plays with git rebasing)

File: tests/migrations/test_commands.py [hunk 1]
```diff
@@ -727,17 +729,34 @@ def test_makemigrations_interactive_accept(self):
...
                 self.fail("Makemigrations failed while running interactive questioner")
             self.assertIn("Created new merge migration", force_text(out.getvalue()))
 
+    def test_makemigrations_default_merge_name(self):
+        test_date = datetime.datetime(2016, 1, 2, 3, 4)
+        test_date_mock = mock.patch('django.utils.timezone.now', mock.Mock(return_value=test_date))
+        with mock.patch('django.db.migrations.questioner.input', mock.Mock(return_value='y')), test_date_mock:
+            out = six.StringIO()
+
+            try:
```
Comment (timgraham) [inline tests/migrations/test_commands.py]: Yes, please rebase the branch and remove the try/fail pattern as done in 6729b96d8a15048b2295c916c5b881a59d9417a0. If you're unfamiliar with the process you might find https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/working-with-git/#rebasing-branches helpful.

File: tests/migrations/test_commands.py [hunk 2]
```diff
@@ -727,17 +729,34 @@ def test_makemigrations_interactive_accept(self):
...
+                    self.assertTrue(os.path.exists(merge_file))
+            except CommandError:
+                self.fail("Makemigrations failed while running interactive questioner")
+            self.assertIn("Created new merge migration", force_text(out.getvalue()))
+
     def test_makemigrations_non_interactive_not_null_addition(self):
         """
         Tests that non-interactive makemigrations fails when a default is missing on a new not-null field.
         """
+
```
Comment (timgraham) [inline tests/migrations/test_commands.py]: Please revert unrelated cosmetic changes to keep the diff clean.

File: tests/migrations/test_commands.py [hunk 3]
```diff
@@ -712,7 +713,8 @@ def test_makemigrations_interactive_reject(self):
         with mock.patch('django.db.migrations.questioner.input', mock.Mock(return_value='N')):
             try:
                 with self.temporary_migration_module(module="migrations.test_migrations_conflict") as migration_dir:
-                    call_command("makemigrations", "migrations", merge=True, interactive=True, verbosity=0)
+                    call_command("makemigrations", "migrations", name="merge", merge=True,
```
Comment (timgraham) [inline tests/migrations/test_commands.py]: If we have to multiline, usually we go with hanging indent like this:
```
call_command(
    'loaddata', ...
    ...
)
```

**Comment by rtpg** [comment]
I cleaned up the code according to the suggestions and rebased it down to one commit, I believe this can be merged now.

**Comment by timgraham** [comment]
Looks mostly good. How about these edits? https://github.com/rtpg/django/pull/1

**Comment by rtpg** [comment]
Those look good, merging it in.
